{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Setup"
      ],
      "metadata": {
        "id": "ZBP3mFgsZTxA"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "This notebook demonstrates how to generate predictions using MHCflurry."
      ],
      "metadata": {
        "id": "c4ukEYkrco5H"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install the package and download models\n",
        "!pip install -q mhcflurry\n",
        "!mhcflurry-downloads --quiet fetch models_class1_presentation"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uUFQLxFgZTAO",
        "outputId": "66ad515b-81bb-46ea-9760-58939109f73a"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[K     |████████████████████████████████| 140 kB 5.0 MB/s \n",
            "\u001b[K     |████████████████████████████████| 103 kB 7.7 MB/s \n",
            "\u001b[K     |████████████████████████████████| 61 kB 275 kB/s \n",
            "\u001b[K     |████████████████████████████████| 636 kB 9.9 MB/s \n",
            "\u001b[K     |████████████████████████████████| 130 kB 12.9 MB/s \n",
            "\u001b[?25h  Building wheel for np-utils (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for serializable (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for typechecks (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "135MB [00:01, 73.1MB/s]               \n",
            "Extracting: 100% 62/62 [00:15<00:00,  3.89it/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Imports\n",
        "import mhcflurry\n",
        "from google.colab import files\n",
        "\n",
        "# Quiet warnings\n",
        "import warnings\n",
        "warnings.filterwarnings('ignore')"
      ],
      "metadata": {
        "id": "1sqAFdItWwd5"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sHCsU4dNRGBk",
        "outputId": "9fd7fec5-ce28-416f-b9b1-aeb4062c79a3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Forcing tensorflow backend.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/compat/v2_compat.py:107: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "non-resource variables are not supported in the long term\n",
            "WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/mhcflurry/common.py:131: The name tf.keras.backend.set_session is deprecated. Please use tf.compat.v1.keras.backend.set_session instead.\n",
            "\n",
            "WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/keras/initializers/initializers_v1.py:278: calling RandomUniform.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Call initializer instance with the dtype argument instead of passing it to the constructor\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Class1PresentationPredictor at 0x7fea8ee35090 [mhcflurry 2.0.6] generated on Thu Jun 11 13:37:18 2020>"
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ],
      "source": [
        "# Load a predictor\n",
        "predictor = mhcflurry.Class1PresentationPredictor.load()\n",
        "predictor"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Predict for specified peptides"
      ],
      "metadata": {
        "id": "SZyoFee7ZlaH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "peptides = \"\"\"\n",
        "NLVPMVATV\n",
        "RANDMPEPTIDE\n",
        "SIINFEKL\n",
        "\"\"\".split()\n",
        "\n",
        "alleles = \"A*02:01 B*27:01 H2-Kb\".split()\n",
        "\n",
        "results1 = predictor.predict(peptides, alleles)\n",
        "results1"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 342
        },
        "id": "LaoQdnXlZLo8",
        "outputId": "755bd5b8-6dd8-4532-8781-395552be82e0"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Predicting processing.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1/1 [00:05<00:00,  5.26s/it]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Predicting affinities.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\r  0%|          | 0/3 [00:00<?, ?it/s]WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/init_ops.py:93: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Call initializer instance with the dtype argument instead of passing it to the constructor\n",
            "WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/init_ops.py:93: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Call initializer instance with the dtype argument instead of passing it to the constructor\n",
            "100%|██████████| 3/3 [00:09<00:00,  3.30s/it]\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        peptide  peptide_num sample_name      affinity best_allele  \\\n",
              "0     NLVPMVATV            0     sample1     16.570972     A*02:01   \n",
              "1  RANDMPEPTIDE            1     sample1  21780.313255     B*27:01   \n",
              "2      SIINFEKL            2     sample1     19.707210       H2-Kb   \n",
              "\n",
              "   processing_score  presentation_score  presentation_percentile  \n",
              "0          0.533008            0.970187                 0.018723  \n",
              "1          0.008492            0.004732                62.744674  \n",
              "2          0.264710            0.914111                 0.099511  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3bcfc280-e3d4-4912-ab50-d4e8595b95b5\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>peptide</th>\n",
              "      <th>peptide_num</th>\n",
              "      <th>sample_name</th>\n",
              "      <th>affinity</th>\n",
              "      <th>best_allele</th>\n",
              "      <th>processing_score</th>\n",
              "      <th>presentation_score</th>\n",
              "      <th>presentation_percentile</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>NLVPMVATV</td>\n",
              "      <td>0</td>\n",
              "      <td>sample1</td>\n",
              "      <td>16.570972</td>\n",
              "      <td>A*02:01</td>\n",
              "      <td>0.533008</td>\n",
              "      <td>0.970187</td>\n",
              "      <td>0.018723</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>RANDMPEPTIDE</td>\n",
              "      <td>1</td>\n",
              "      <td>sample1</td>\n",
              "      <td>21780.313255</td>\n",
              "      <td>B*27:01</td>\n",
              "      <td>0.008492</td>\n",
              "      <td>0.004732</td>\n",
              "      <td>62.744674</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SIINFEKL</td>\n",
              "      <td>2</td>\n",
              "      <td>sample1</td>\n",
              "      <td>19.707210</td>\n",
              "      <td>H2-Kb</td>\n",
              "      <td>0.264710</td>\n",
              "      <td>0.914111</td>\n",
              "      <td>0.099511</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3bcfc280-e3d4-4912-ab50-d4e8595b95b5')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3bcfc280-e3d4-4912-ab50-d4e8595b95b5 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3bcfc280-e3d4-4912-ab50-d4e8595b95b5');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download results\n",
        "results1.to_csv('mhcflurry-results.csv')\n",
        "files.download('mhcflurry-results.csv')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "B2YXOk6waeli",
        "outputId": "c933a1b7-65c5-44da-dacc-8ab3813de681"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_044dbb46-ccba-49fe-876b-f71d363c8833\", \"mhcflurry-results.csv\", 434)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# See help for more options:\n",
        "help(predictor.predict)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OIYIyhiudmis",
        "outputId": "d438606b-e747-4f28-9bdc-bb41733746bb"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Help on method predict in module mhcflurry.class1_presentation_predictor:\n",
            "\n",
            "predict(peptides, alleles, sample_names=None, n_flanks=None, c_flanks=None, include_affinity_percentile=False, verbose=1, throw=True) method of mhcflurry.class1_presentation_predictor.Class1PresentationPredictor instance\n",
            "    Predict presentation scores across a set of peptides.\n",
            "    \n",
            "    Presentation scores combine predictions for MHC I binding affinity\n",
            "    and antigen processing.\n",
            "    \n",
            "    This method returns a pandas.DataFrame giving presentation scores plus\n",
            "    the binding affinity and processing predictions and other intermediate\n",
            "    results.\n",
            "    \n",
            "    Example:\n",
            "    \n",
            "    >>> predictor = Class1PresentationPredictor.load()\n",
            "    >>> predictor.predict(\n",
            "    ...    peptides=[\"SIINFEKL\", \"PEPTIDE\"],\n",
            "    ...    n_flanks=[\"NNN\", \"SNS\"],\n",
            "    ...    c_flanks=[\"CCC\", \"CNC\"],\n",
            "    ...    alleles={\n",
            "    ...        \"sample1\": [\"A0201\", \"A0301\", \"B0702\"],\n",
            "    ...        \"sample2\": [\"A0101\", \"C0202\"],\n",
            "    ...    },\n",
            "    ...    verbose=0)\n",
            "        peptide n_flank c_flank  peptide_num sample_name   affinity best_allele  processing_score  presentation_score  presentation_percentile\n",
            "    0  SIINFEKL     NNN     CCC            0     sample1  11927.161       A0201             0.838               0.145                    2.282\n",
            "    1   PEPTIDE     SNS     CNC            1     sample1  32507.082       A0201             0.025               0.003                  100.000\n",
            "    2  SIINFEKL     NNN     CCC            0     sample2   2725.593       C0202             0.838               0.416                    1.017\n",
            "    3   PEPTIDE     SNS     CNC            1     sample2  28304.338       C0202             0.025               0.003                   99.287\n",
            "    \n",
            "    You can also specify sample_names, in which case peptide is evaluated\n",
            "    for binding the alleles in the corresponding sample only. See\n",
            "    `predict_affinity` for an examples.\n",
            "    \n",
            "    Parameters\n",
            "    ----------\n",
            "    peptides : list of string\n",
            "        Peptide sequences\n",
            "    alleles : list of string or dict of string -> list of string\n",
            "        If you are predicting for a single sample, pass a list of strings\n",
            "        (up to 6) indicating the genotype. If you are predicting across\n",
            "        multiple samples, pass a dict where the keys are (arbitrary)\n",
            "        sample names and the values are the alleles to predict for that\n",
            "        sample. Set to an empty list or dict to perform processing\n",
            "        prediction only.\n",
            "    sample_names : list of string [same length as peptides]\n",
            "        If you are passing a dict for 'alleles', you can use this\n",
            "        argument to specify which peptides go with which samples. If it is\n",
            "        None, then predictions will be performed for each peptide across all\n",
            "        samples.\n",
            "    n_flanks : list of string [same length as peptides]\n",
            "        Upstream sequences before the peptide. Sequences of any length can\n",
            "        be given and a suffix of the size supported by the model will be\n",
            "        used.\n",
            "    c_flanks : list of string [same length as peptides]\n",
            "        Downstream sequences after the peptide. Sequences of any length can\n",
            "        be given and a prefix of the size supported by the model will be\n",
            "        used.\n",
            "    include_affinity_percentile : bool\n",
            "        Whether to include affinity percentile ranks\n",
            "    verbose : int\n",
            "        Set to 0 for quiet.\n",
            "    throw : verbose\n",
            "        Whether to throw exception (vs. just log a warning) on invalid\n",
            "        peptides, etc.\n",
            "    \n",
            "    Returns\n",
            "    -------\n",
            "    pandas.DataFrame\n",
            "    \n",
            "    Presentation scores and intermediate results.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Predict by scanning across protein sequences"
      ],
      "metadata": {
        "id": "TAJlyQ-4axay"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Paste your fasta here\n",
        "proteins_fasta = \"\"\"\n",
        ">tr|A0A6B9WFC7|A0A6B9WFC7_SARS2 Envelope small membrane protein\n",
        "MYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNIVNVSLVKPSFYVYS\n",
        "RVKNLNSSRVPDLLV\n",
        ">tr|A0A6B9W0L4|A0A6B9W0L4_SARS2 ORF6 protein\n",
        "MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTENKYSQLDEEQPMEI\n",
        "D\n",
        ">tr|A0A6G7S6S0|A0A6G7S6S0_SARS2 Nonstructural protein NS3\n",
        "MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWLIVGVALLAVFQSAS\n",
        "KIITLKKRWQLALSKGVHFVCNLLLLFVTVYSHLLLVAAGLEAPFLYLYALVYFLQSINF\n",
        "VRIIMRLWLCWKCRSKNPLLYDANYFLCWHTNCYDYCIPYNSVTSSIVITSGDGTTSPIS\n",
        "EHDYQIGGYTEKWESGVKDCVVLHSYFTSDYYQLYSTQLSTDTGVEHVTFFIYNKIVDEP\n",
        "EEHVQIHTIDGSSGVVNPVMEPIYDEPTTTTSVPL\n",
        ">tr|A0A6B9VLF3|A0A6B9VLF3_SARS2 Membrane protein\n",
        "MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFLYIIKLIFLWLLWPV\n",
        "TLACFVLAAVYRINWITGGIAIAMACLVGLMWLSYFIASFRLFARTRSMWSFNPETNILL\n",
        "NVPLHGTILTRPLLESELVIGAVILRGHLRIAGHHLGRCDIKDLPKEITVATSRTLSYYK\n",
        "LGASQRVAGDSGFAAYSRYRIGNYKLNTDHSSSSDNIALLVQ\n",
        "\"\"\"\n",
        "\n",
        "import mhcflurry.fasta\n",
        "\n",
        "with open(\"temp.fa\", \"w\") as fd:\n",
        "    fd.write(proteins_fasta)\n",
        "\n",
        "proteins = mhcflurry.fasta.read_fasta_to_dataframe(\"temp.fa\").set_index(\"sequence_id\")\n",
        "proteins"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "IORcEbHkbBsJ",
        "outputId": "db4bb0bd-0602-4452-dd4e-55dbfa7cb90d"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                                         sequence\n",
              "sequence_id                                                                      \n",
              "tr|A0A6B9WFC7|A0A6B9WFC7_SARS2  MYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...\n",
              "tr|A0A6B9W0L4|A0A6B9W0L4_SARS2  MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...\n",
              "tr|A0A6G7S6S0|A0A6G7S6S0_SARS2  MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...\n",
              "tr|A0A6B9VLF3|A0A6B9VLF3_SARS2  MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL..."
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-62abe4e0-7cd3-4f1e-84eb-93cae45f41a3\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sequence</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sequence_id</th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>tr|A0A6B9WFC7|A0A6B9WFC7_SARS2</th>\n",
              "      <td>MYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>tr|A0A6B9W0L4|A0A6B9W0L4_SARS2</th>\n",
              "      <td>MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>tr|A0A6G7S6S0|A0A6G7S6S0_SARS2</th>\n",
              "      <td>MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>tr|A0A6B9VLF3|A0A6B9VLF3_SARS2</th>\n",
              "      <td>MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-62abe4e0-7cd3-4f1e-84eb-93cae45f41a3')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-62abe4e0-7cd3-4f1e-84eb-93cae45f41a3 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-62abe4e0-7cd3-4f1e-84eb-93cae45f41a3');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Define alleles for each sample\n",
        "alleles={\n",
        "    \"my-sample\": [\"A0201\", \"A0301\", \"B0702\", \"C0802\"],\n",
        "}"
      ],
      "metadata": {
        "id": "zwPHH09RcgCt"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 559
        },
        "id": "tYRpxn5YRGBk",
        "outputId": "7b07a064-fcdf-4fec-d05a-242fdda2a6ad"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Predicting processing.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1/1 [00:13<00:00, 13.82s/it]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Predicting affinities.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 4/4 [00:06<00:00,  1.62s/it]\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                      sequence_name  pos     peptide n_flank c_flank  \\\n",
              "0    tr|A0A6G7S6S0|A0A6G7S6S0_SARS2  138   LLYDANYFL   RSKNP   CWHTN   \n",
              "1    tr|A0A6G7S6S0|A0A6G7S6S0_SARS2  106   YLYALVYFL   EAPFL   QSINF   \n",
              "2    tr|A0A6G7S6S0|A0A6G7S6S0_SARS2   71   ALSKGVHFV   KRWQL   CNLLL   \n",
              "3    tr|A0A6B9WFC7|A0A6B9WFC7_SARS2   49   SLVKPSFYV   NIVNV   YSRVK   \n",
              "4    tr|A0A6B9WFC7|A0A6B9WFC7_SARS2   19   FLAFVVFLL   NSVLL   VTLAI   \n",
              "..                              ...  ...         ...     ...     ...   \n",
              "188  tr|A0A6B9WFC7|A0A6B9WFC7_SARS2   15  SVLLFLAFVV   TLIVN   FLLVT   \n",
              "189  tr|A0A6G7S6S0|A0A6G7S6S0_SARS2   57    SASKIITL   LAVFQ   KKRWQ   \n",
              "190  tr|A0A6G7S6S0|A0A6G7S6S0_SARS2  169  TSGDGTTSPI   SSIVI   SEHDY   \n",
              "191  tr|A0A6B9VLF3|A0A6B9VLF3_SARS2   71   RINWITGGI   LAAVY   AIAMA   \n",
              "192  tr|A0A6G7S6S0|A0A6G7S6S0_SARS2    1   DLFMRIFTI       M   GTVTL   \n",
              "\n",
              "    sample_name    affinity best_allele  affinity_percentile  \\\n",
              "0     my-sample   10.659104       A0201             0.003625   \n",
              "1     my-sample   11.053785       A0201             0.006750   \n",
              "2     my-sample   11.501204       A0201             0.011500   \n",
              "3     my-sample   11.930823       A0201             0.013500   \n",
              "4     my-sample   12.318483       A0201             0.015875   \n",
              "..          ...         ...         ...                  ...   \n",
              "188   my-sample  466.913027       A0201             1.297625   \n",
              "189   my-sample  471.300226       C0802             0.774375   \n",
              "190   my-sample  473.865753       C0802             0.774375   \n",
              "191   my-sample  475.852826       A0201             1.306500   \n",
              "192   my-sample  498.907589       A0201             1.329875   \n",
              "\n",
              "     processing_score  presentation_score  presentation_percentile  \n",
              "0            0.157175            0.921852                 0.088804  \n",
              "1            0.014756            0.868851                 0.171848  \n",
              "2            0.676803            0.987502                 0.002065  \n",
              "3            0.091771            0.891807                 0.135353  \n",
              "4            0.007210            0.852791                 0.196277  \n",
              "..                ...                 ...                      ...  \n",
              "188          0.010083            0.145307                 2.271005  \n",
              "189          0.772850            0.753873                 0.351359  \n",
              "190          0.000247            0.138992                 2.345462  \n",
              "191          0.166255            0.232094                 1.656413  \n",
              "192          0.972156            0.860781                 0.184022  \n",
              "\n",
              "[193 rows x 12 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5f2fb820-3c51-4916-ad31-6354fba87d6a\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sequence_name</th>\n",
              "      <th>pos</th>\n",
              "      <th>peptide</th>\n",
              "      <th>n_flank</th>\n",
              "      <th>c_flank</th>\n",
              "      <th>sample_name</th>\n",
              "      <th>affinity</th>\n",
              "      <th>best_allele</th>\n",
              "      <th>affinity_percentile</th>\n",
              "      <th>processing_score</th>\n",
              "      <th>presentation_score</th>\n",
              "      <th>presentation_percentile</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>tr|A0A6G7S6S0|A0A6G7S6S0_SARS2</td>\n",
              "      <td>138</td>\n",
              "      <td>LLYDANYFL</td>\n",
              "      <td>RSKNP</td>\n",
              "      <td>CWHTN</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>10.659104</td>\n",
              "      <td>A0201</td>\n",
              "      <td>0.003625</td>\n",
              "      <td>0.157175</td>\n",
              "      <td>0.921852</td>\n",
              "      <td>0.088804</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>tr|A0A6G7S6S0|A0A6G7S6S0_SARS2</td>\n",
              "      <td>106</td>\n",
              "      <td>YLYALVYFL</td>\n",
              "      <td>EAPFL</td>\n",
              "      <td>QSINF</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>11.053785</td>\n",
              "      <td>A0201</td>\n",
              "      <td>0.006750</td>\n",
              "      <td>0.014756</td>\n",
              "      <td>0.868851</td>\n",
              "      <td>0.171848</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>tr|A0A6G7S6S0|A0A6G7S6S0_SARS2</td>\n",
              "      <td>71</td>\n",
              "      <td>ALSKGVHFV</td>\n",
              "      <td>KRWQL</td>\n",
              "      <td>CNLLL</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>11.501204</td>\n",
              "      <td>A0201</td>\n",
              "      <td>0.011500</td>\n",
              "      <td>0.676803</td>\n",
              "      <td>0.987502</td>\n",
              "      <td>0.002065</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>tr|A0A6B9WFC7|A0A6B9WFC7_SARS2</td>\n",
              "      <td>49</td>\n",
              "      <td>SLVKPSFYV</td>\n",
              "      <td>NIVNV</td>\n",
              "      <td>YSRVK</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>11.930823</td>\n",
              "      <td>A0201</td>\n",
              "      <td>0.013500</td>\n",
              "      <td>0.091771</td>\n",
              "      <td>0.891807</td>\n",
              "      <td>0.135353</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>tr|A0A6B9WFC7|A0A6B9WFC7_SARS2</td>\n",
              "      <td>19</td>\n",
              "      <td>FLAFVVFLL</td>\n",
              "      <td>NSVLL</td>\n",
              "      <td>VTLAI</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>12.318483</td>\n",
              "      <td>A0201</td>\n",
              "      <td>0.015875</td>\n",
              "      <td>0.007210</td>\n",
              "      <td>0.852791</td>\n",
              "      <td>0.196277</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>188</th>\n",
              "      <td>tr|A0A6B9WFC7|A0A6B9WFC7_SARS2</td>\n",
              "      <td>15</td>\n",
              "      <td>SVLLFLAFVV</td>\n",
              "      <td>TLIVN</td>\n",
              "      <td>FLLVT</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>466.913027</td>\n",
              "      <td>A0201</td>\n",
              "      <td>1.297625</td>\n",
              "      <td>0.010083</td>\n",
              "      <td>0.145307</td>\n",
              "      <td>2.271005</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>189</th>\n",
              "      <td>tr|A0A6G7S6S0|A0A6G7S6S0_SARS2</td>\n",
              "      <td>57</td>\n",
              "      <td>SASKIITL</td>\n",
              "      <td>LAVFQ</td>\n",
              "      <td>KKRWQ</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>471.300226</td>\n",
              "      <td>C0802</td>\n",
              "      <td>0.774375</td>\n",
              "      <td>0.772850</td>\n",
              "      <td>0.753873</td>\n",
              "      <td>0.351359</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>190</th>\n",
              "      <td>tr|A0A6G7S6S0|A0A6G7S6S0_SARS2</td>\n",
              "      <td>169</td>\n",
              "      <td>TSGDGTTSPI</td>\n",
              "      <td>SSIVI</td>\n",
              "      <td>SEHDY</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>473.865753</td>\n",
              "      <td>C0802</td>\n",
              "      <td>0.774375</td>\n",
              "      <td>0.000247</td>\n",
              "      <td>0.138992</td>\n",
              "      <td>2.345462</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>191</th>\n",
              "      <td>tr|A0A6B9VLF3|A0A6B9VLF3_SARS2</td>\n",
              "      <td>71</td>\n",
              "      <td>RINWITGGI</td>\n",
              "      <td>LAAVY</td>\n",
              "      <td>AIAMA</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>475.852826</td>\n",
              "      <td>A0201</td>\n",
              "      <td>1.306500</td>\n",
              "      <td>0.166255</td>\n",
              "      <td>0.232094</td>\n",
              "      <td>1.656413</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>192</th>\n",
              "      <td>tr|A0A6G7S6S0|A0A6G7S6S0_SARS2</td>\n",
              "      <td>1</td>\n",
              "      <td>DLFMRIFTI</td>\n",
              "      <td>M</td>\n",
              "      <td>GTVTL</td>\n",
              "      <td>my-sample</td>\n",
              "      <td>498.907589</td>\n",
              "      <td>A0201</td>\n",
              "      <td>1.329875</td>\n",
              "      <td>0.972156</td>\n",
              "      <td>0.860781</td>\n",
              "      <td>0.184022</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>193 rows × 12 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5f2fb820-3c51-4916-ad31-6354fba87d6a')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-5f2fb820-3c51-4916-ad31-6354fba87d6a button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-5f2fb820-3c51-4916-ad31-6354fba87d6a');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ],
      "source": [
        "# Predict across protein sequences and return peptides with predicted affinity\n",
        "# less than 500 nM.\n",
        "results2 = predictor.predict_sequences(\n",
        "    sequences=proteins.sequence.to_dict(),\n",
        "    alleles=alleles,\n",
        "    result=\"filtered\",\n",
        "    comparison_quantity=\"affinity\",\n",
        "    filter_value=500)\n",
        "results2"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download results\n",
        "results2.to_csv('mhcflurry-results.csv')\n",
        "files.download('mhcflurry-results.csv')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "CASrg0X5dgon",
        "outputId": "8d333984-00cc-44d2-f5f4-6308fd7348a0"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_b220af07-7880-4f63-93dd-8780f762a0a9\", \"mhcflurry-results.csv\", 32592)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# See help for more options:\n",
        "help(predictor.predict_sequences)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "RtbpRsdTdMRL",
        "outputId": "dda97aff-cdb7-4d8c-d950-df13a7aecbd6"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Help on method predict_sequences in module mhcflurry.class1_presentation_predictor:\n",
            "\n",
            "predict_sequences(sequences, alleles, result='best', comparison_quantity=None, filter_value=None, peptide_lengths=(8, 9, 10, 11), use_flanks=True, include_affinity_percentile=True, verbose=1, throw=True) method of mhcflurry.class1_presentation_predictor.Class1PresentationPredictor instance\n",
            "    Predict presentation across protein sequences.\n",
            "    \n",
            "    Example:\n",
            "    \n",
            "    >>> predictor = Class1PresentationPredictor.load()\n",
            "    >>> predictor.predict_sequences(\n",
            "    ...    sequences={\n",
            "    ...        'protein1': \"MDSKGSSQKGSRLLLLLVVSNLL\",\n",
            "    ...        'protein2': \"SSLPTPEDKEQAQQTHH\",\n",
            "    ...    },\n",
            "    ...    alleles={\n",
            "    ...        \"sample1\": [\"A0201\", \"A0301\", \"B0702\"],\n",
            "    ...        \"sample2\": [\"A0101\", \"C0202\"],\n",
            "    ...    },\n",
            "    ...    result=\"filtered\",\n",
            "    ...    comparison_quantity=\"affinity\",\n",
            "    ...    filter_value=500,\n",
            "    ...    verbose=0)\n",
            "      sequence_name  pos     peptide n_flank c_flank sample_name  affinity best_allele  affinity_percentile  processing_score  presentation_score  presentation_percentile\n",
            "    0      protein1   14   LLLVVSNLL   GSRLL             sample1    57.180       A0201                0.398             0.233               0.754                    0.351\n",
            "    1      protein1   13   LLLLVVSNL   KGSRL       L     sample1    57.339       A0201                0.398             0.031               0.586                    0.643\n",
            "    2      protein1    5   SSQKGSRLL   MDSKG   LLLVV     sample2   110.779       C0202                0.782             0.061               0.456                    0.920\n",
            "    3      protein1    6   SQKGSRLLL   DSKGS   LLVVS     sample2   254.480       C0202                1.735             0.102               0.303                    1.356\n",
            "    4      protein1   13  LLLLVVSNLL   KGSRL             sample1   260.390       A0201                1.012             0.158               0.345                    1.215\n",
            "    5      protein1   12  LLLLLVVSNL   QKGSR       L     sample1   308.150       A0201                1.094             0.015               0.206                    1.802\n",
            "    6      protein2    0   SSLPTPEDK           EQAQQ     sample2   410.354       C0202                2.398             0.003               0.158                    2.155\n",
            "    7      protein1    5    SSQKGSRL   MDSKG   LLLLV     sample2   444.321       C0202                2.512             0.026               0.159                    2.138\n",
            "    8      protein2    0   SSLPTPEDK           EQAQQ     sample1   459.296       A0301                0.971             0.003               0.144                    2.292\n",
            "    9      protein1    4   GSSQKGSRL    MDSK   LLLLV     sample2   469.052       C0202                2.595             0.014               0.146                    2.261\n",
            "    \n",
            "    Parameters\n",
            "    ----------\n",
            "    sequences : str, list of string, or string -> string dict\n",
            "        Protein sequences. If a dict is given, the keys are arbitrary (\n",
            "        e.g. protein names), and the values are the amino acid sequences.\n",
            "    alleles : list of string, list of list of string, or dict of string -> list of string\n",
            "        MHC I alleles. Can be: (1) a string (a single allele), (2) a list of\n",
            "        strings (a single genotype), (3) a list of list of strings\n",
            "        (multiple genotypes, where the total number of genotypes must equal\n",
            "        the number of sequences), or (4) a dict giving multiple genotypes,\n",
            "        which will each be run over the sequences.\n",
            "    result : string\n",
            "        Specify 'best' to return the strongest peptide for each sequence,\n",
            "        'all' to return predictions for all peptides, or 'filtered' to\n",
            "        return predictions where the comparison_quantity is stronger\n",
            "        (i.e (<) for affinity, (>) for scores) than filter_value.\n",
            "    comparison_quantity : string\n",
            "        One of \"presentation_score\", \"processing_score\", \"affinity\", or\n",
            "        \"affinity_percentile\". Prediction to use to rank (if result is\n",
            "        \"best\") or filter (if result is \"filtered\") results. Default is\n",
            "        \"presentation_score\".\n",
            "    filter_value : float\n",
            "        Threshold value to use, only relevant when result is \"filtered\".\n",
            "        If comparison_quantity is \"affinity\", then all results less than\n",
            "        (i.e. tighter than) the specified nM affinity are retained. If it's\n",
            "        \"presentation_score\" or \"processing_score\" then results greater than\n",
            "        the indicated filter_value are retained.\n",
            "    peptide_lengths : list of int\n",
            "        Peptide lengths to predict for.\n",
            "    use_flanks : bool\n",
            "        Whether to include flanking sequences when running the AP predictor\n",
            "        (for better cleavage prediction).\n",
            "    include_affinity_percentile : bool\n",
            "        Whether to include affinity percentile ranks in output.\n",
            "    verbose : int\n",
            "        Set to 0 for quiet mode.\n",
            "    throw : boolean\n",
            "        Whether to throw exceptions (vs. log warnings) on invalid inputs.\n",
            "    \n",
            "    Returns\n",
            "    -------\n",
            "    pandas.DataFrame with columns:\n",
            "        peptide, n_flank, c_flank, sequence_name, affinity, best_allele,\n",
            "        processing_score, presentation_score\n",
            "\n"
          ]
        }
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.6.1"
    },
    "colab": {
      "name": "mhcflurry-colab.ipynb",
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
